nupic/network/synapse.py

```diff
@@ -0,0 +1,38 @@
+﻿# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2014, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+
+
+class Synapse(object):

```

Comment:
If I replace `obj = objects.setdefault(index, MutableObject("bar"))` with `objects[index] = MutableObject("bar")`, it shaves ~2 seconds off the total, bringing execution down to 13.204 seconds.  In a realistic caching scenario, you would need to do some sort of existence check anyway, hence the `setdefault()`.

```
Austins-MacBook-Pro:Desktop amarshall$ python -m cProfile mutant.py mutate
         10000005 function calls in 13.204 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   13.204   13.204 mutant.py:1(<module>)
 10000000    6.351    0.000    6.351    0.000 mutant.py:10(__init__)
        1    0.000    0.000    0.000    0.000 mutant.py:15(SlottedMutableObject)
        1    6.853    6.853   13.203   13.203 mutant.py:26(mutate)
        1    0.000    0.000    0.000    0.000 mutant.py:9(MutableObject)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```